File: Packages/WhoopStore/Sources/WhoopStore/StrengthWorkoutPlanning.swift
import Foundation

public enum StrengthProgressionReason: String, Codable, Sendable {
    case firstSession
    case repeatLoad
    case repRangeAdvanced
    case linearAdvanced
    case timeAdvanced
    case bodyweightRepProgress
}

public struct StrengthPlannedSet: Equatable, Sendable {
    public let setType: String
    public let reps: Int?
    public let loadKg: Double?
    public let durationS: Int?
    public let restSecondsAfter: Int?
}

public struct StrengthWorkoutPrescription: Equatable, Sendable {
    public let sets: [StrengthPlannedSet]
    public let reason: StrengthProgressionReason
    public let previousSessionAt: Int?
}

public struct StrengthEstimatedMaximum: Equatable, Sendable {
    public let exerciseId: String
    public let kilograms: Double
    public let sourceLoadKg: Double
    public let sourceReps: Int
    public let sessionId: String
    public let recordedAt: Int
}

public enum StrengthWorkoutPlanner {
    public static func prescription(
        exercise: StrengthExerciseRow,
        prescription: StrengthRoutineExerciseRow,
        history: [StrengthSessionSnapshot]
    ) -> StrengthWorkoutPrescription {
        let plan = StrengthTrainingContract.exercisePlan(from: prescription.planJSON)
        let previous = history
            .filter { $0.session.endedAt != nil }
            .sorted {
                if $0.session.startedAt != $1.session.startedAt {
                    return $0.session.startedAt > $1.session.startedAt
                }
                return $0.session.id < $1.session.id
            }
            .first { snapshot in
                snapshot.sets.contains {
                    $0.exerciseId == exercise.id
                        && $0.completedAt != nil
                        && progressionSetTypes.contains($0.setType)
                }
            }
        let previousSets = previous?.sets
            .filter {
                $0.exerciseId == exercise.id
                    && $0.completedAt != nil
                    && progressionSetTypes.contains($0.setType)
            }
            .sorted { $0.setPosition < $1.setPosition } ?? []

        let lower = max(1, prescription.targetRepsMin ?? 8)
        let upper = max(lower, prescription.targetRepsMax ?? lower)
        let priorLoad = previousSets.compactMap(\.loadKg).max()
        var workLoad = priorLoad ?? plan.targetLoadKg
        let compared = Array(previousSets.prefix(prescription.targetSets))
        var workRepTargets = (0..<prescription.targetSets).map { index in
            guard compared.indices.contains(index), let reps = compared[index].reps else {
                return lower
            }
            return min(max(reps, lower), upper)
        }
        var duration = compared.compactMap(\.durationS).max() ?? plan.targetDurationS
        var reason: StrengthProgressionReason = previous == nil ? .firstSession : .repeatLoad

        switch plan.progression {
        case "double_progression":
            if !compared.isEmpty,
                compared.count >= prescription.targetSets,
               compared.allSatisfy({ ($0.reps ?? 0) >= upper }) {
                if let current = workLoad {
                    if let advanced = progressedLoad(
                        current: current,
                        configuredStep: plan.loadStepKg
                    ) {
                        workLoad = advanced
                        workRepTargets = Array(repeating: lower, count: prescription.targetSets)
                        reason = .repRangeAdvanced
                    }
                } else {
                    let step = plan.repsPerSide ? 2 : 1
                    let completedFloor = compared.compactMap(\.reps).min() ?? upper
                    let advanced = min(
                        StrengthTrainingContract.maxReps,
                        max(upper, completedFloor) + step
                    )
                    workRepTargets = Array(
                        repeating: advanced,
                        count: prescription.targetSets
                    )
                    if advanced > completedFloor { reason = .bodyweightRepProgress }
                }
            } else if !compared.isEmpty {
                let step = plan.repsPerSide ? 2 : 1
                workRepTargets = (0..<prescription.targetSets).map { index in
                    guard compared.indices.contains(index), let reps = compared[index].reps else {
                        return lower
                    }
                    return min(upper, max(lower, reps) + step)
                }
            }
        case "linear":
            if !compared.isEmpty,
               compared.count >= prescription.targetSets,
               compared.allSatisfy({ ($0.reps ?? 0) >= lower }),
               let current = workLoad {
                if let advanced = progressedLoad(
                    current: current,
                    configuredStep: plan.loadStepKg
                ) {
                    workLoad = advanced
                    reason = .linearAdvanced
                }
            }
        case "time":
            let target = max(
                plan.targetDurationS ?? 30,
                compared.compactMap(\.durationS).max() ?? 0
            )
            if !compared.isEmpty,
               compared.count >= prescription.targetSets,
               compared.allSatisfy({ ($0.durationS ?? 0) >= target }) {
                let advanced = min(StrengthTrainingContract.maxDurationSeconds, target + 5)
                duration = advanced
                if advanced > target { reason = .timeAdvanced }
            } else {
                duration = target
            }
        default:
            break
        }

        if exercise.equipment == "bodyweight", plan.targetLoadKg == nil, priorLoad == nil {
            workLoad = nil
        }
        let isTimed = plan.mode == "timed" || exercise.movementPattern == "cardio"
        let warmups = warmupRows(
            count: isTimed ? 0 : plan.warmupSets,
            workLoad: workLoad,
            workReps: workRepTargets.first ?? lower
        )
        var work = (0..<prescription.targetSets).map { index in
            return StrengthPlannedSet(
                setType: exercise.equipment == "bodyweight" && workLoad == nil
                    ? "bodyweight"
                    : "working",
                reps: isTimed ? nil : workRepTargets[index],
                loadKg: workLoad,
                durationS: isTimed ? (duration ?? 30) : nil,
                restSecondsAfter: nil
            )
        }
        if !isTimed, plan.setStyle == "drop", let workLoad {
            let dropLoad = roundedLoad(workLoad * (1 - Double(plan.dropPercent) / 100))
            if !work.isEmpty, dropLoad > 0, dropLoad < workLoad {
                work[work.count - 1] = StrengthPlannedSet(
                    setType: work[work.count - 1].setType,
                    reps: work[work.count - 1].reps,
                    loadKg: work[work.count - 1].loadKg,
                    durationS: nil,
                    restSecondsAfter: 0
                )
                work.append(StrengthPlannedSet(
                    setType: "drop",
                    reps: workRepTargets.last ?? lower,
                    loadKg: dropLoad,
                    durationS: nil,
                    restSecondsAfter: nil
                ))
            }
        } else if !isTimed, plan.setStyle == "rest_pause", !work.isEmpty {
            work[work.count - 1] = StrengthPlannedSet(
                setType: work[work.count - 1].setType,
                reps: work[work.count - 1].reps,
                loadKg: work[work.count - 1].loadKg,
                durationS: nil,
                restSecondsAfter: plan.restPauseSeconds
            )
            work.append(StrengthPlannedSet(
                setType: "rest_pause",
                reps: nil,
                loadKg: workLoad,
                durationS: nil,
                restSecondsAfter: nil
            ))
        }
        return StrengthWorkoutPrescription(
            sets: warmups + work,
            reason: reason,
            previousSessionAt: previous?.session.startedAt
        )
    }

    /// Resolve the timer after a planned set. Warm-ups keep their own prescription rest, while
    /// working sets inside a superset wait only after the last movement in the round.
    public static func resolvedRestSeconds(
        for target: StrengthPlannedSet,
        prescriptionRestSeconds: Int,
        continuesSuperset: Bool
    ) -> Int {
        if target.setType == "warmup" { return prescriptionRestSeconds }
        return target.restSecondsAfter ?? (continuesSuperset ? 0 : prescriptionRestSeconds)
    }

    /// Epley estimate, bounded to low-repetition completed work where the estimate remains useful.
    public static func estimatedOneRepMaximum(loadKg: Double, reps: Int) -> Double? {
        guard loadKg.isFinite, loadKg > 0, reps >= 1, reps <= 10 else { return nil }
        if reps == 1 { return loadKg }
        return loadKg * (1 + Double(reps) / 30)
    }

    public static func bestEstimatedMaximum(
        exerciseId: String,
        sessions: [StrengthSessionSnapshot]
    ) -> StrengthEstimatedMaximum? {
        sessions
            .filter { $0.session.endedAt != nil }
            .flatMap { snapshot in
                snapshot.sets.compactMap { set -> StrengthEstimatedMaximum? in
                    guard set.exerciseId == exerciseId, set.completedAt != nil,
                          ["working", "failure"].contains(set.setType),
                          let load = set.loadKg, let reps = set.reps,
                          let estimate = estimatedOneRepMaximum(loadKg: load, reps: reps)
                    else { return nil }
                    return StrengthEstimatedMaximum(
                        exerciseId: exerciseId,
                        kilograms: estimate,
                        sourceLoadKg: load,
                        sourceReps: reps,
                        sessionId: snapshot.session.id,
                        recordedAt: snapshot.session.startedAt
                    )
                }
            }
            .max {
                if $0.kilograms != $1.kilograms { return $0.kilograms < $1.kilograms }
                return $0.recordedAt < $1.recordedAt
            }
    }

    private static func warmupRows(
        count: Int,
        workLoad: Double?,
        workReps: Int
    ) -> [StrengthPlannedSet] {
        guard count > 0, let workLoad else { return [] }
        let ladders: [Int: [(loadFraction: Double, reps: Int)]] = [
            1: [(0.6, 5)],
            2: [(0.5, 5), (0.75, 3)],
            3: [(0.4, 5), (0.6, 3), (0.8, 2)],
            4: [(0.35, 5), (0.5, 4), (0.65, 3), (0.8, 2)],
            5: [(0.3, 5), (0.45, 4), (0.6, 3), (0.72, 2), (0.84, 1)],
        ]
        return (ladders[min(5, count)] ?? []).compactMap { step in
            let load = roundedLoad(workLoad * step.loadFraction)
            guard load > 0, load < workLoad else { return nil }
            return StrengthPlannedSet(
                setType: "warmup",
                reps: max(1, min(workReps, step.reps)),
                loadKg: load,
                durationS: nil,
                restSecondsAfter: nil
            )
        }
    }

    private static func roundedLoad(_ kilograms: Double) -> Double {
        (kilograms * 2).rounded() / 2
    }

    private static let progressionSetTypes: Set<String> = ["working", "failure", "bodyweight"]

    /// ACSM recommends a 2–10% load increase after the target is exceeded. Respect the configured
    /// increment while capping it at 10%; if the available 0.5 kg resolution would exceed that cap,
    /// hold the current load instead of inventing a larger jump.
    private static func progressedLoad(current: Double, configuredStep: Double) -> Double? {
        guard current.isFinite, current > 0, configuredStep.isFinite, configuredStep > 0 else {
            return nil
        }
        let maximumIncrease = current * 0.10
        let unrounded = current + min(configuredStep, maximumIncrease)
        var candidate = roundedLoad(unrounded)
        if candidate - current > maximumIncrease + 0.000_001 {
            candidate = floor(unrounded * 2) / 2
        }
        return candidate > current ? candidate : nil
    }

}

File: android/app/src/main/java/com/noop/data/StrengthWorkoutPlanning.kt
package com.noop.data

import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min

enum class StrengthProgressionReason {
    FIRST_SESSION,
    REPEAT_LOAD,
    REP_RANGE_ADVANCED,
    LINEAR_ADVANCED,
    TIME_ADVANCED,
    BODYWEIGHT_REP_PROGRESS,
}

data class StrengthPlannedSet(
    val setType: String,
    val reps: Int?,
    val loadKg: Double?,
    val durationS: Int?,
    val restSecondsAfter: Int?,
)

data class StrengthWorkoutPrescription(
    val sets: List<StrengthPlannedSet>,
    val reason: StrengthProgressionReason,
    val previousSessionAt: Long?,
)

data class StrengthEstimatedMaximum(
    val exerciseId: String,
    val kilograms: Double,
    val sourceLoadKg: Double,
    val sourceReps: Int,
    val sessionId: String,
    val recordedAt: Long,
)

object StrengthWorkoutPlanner {
    fun prescription(
        exercise: StrengthExerciseRow,
        prescription: StrengthRoutineExerciseRow,
        history: List<StrengthSessionSnapshot>,
    ): StrengthWorkoutPrescription {
        val plan = StrengthTrainingContract.exercisePlan(prescription.planJSON)
        val previous = history
            .filter { it.session.endedAt != null }
            .sortedWith(compareByDescending<StrengthSessionSnapshot> { it.session.startedAt }.thenBy { it.session.id })
            .firstOrNull { snapshot ->
                snapshot.sets.any {
                    it.exerciseId == exercise.id &&
                        it.completedAt != null &&
                        it.setType in PROGRESSION_SET_TYPES
                }
            }
        val previousSets = previous?.sets.orEmpty()
            .filter {
                it.exerciseId == exercise.id &&
                    it.completedAt != null &&
                    it.setType in PROGRESSION_SET_TYPES
            }
            .sortedBy { it.setPosition }
        val lower = max(1, prescription.targetRepsMin ?: 8)
        val upper = max(lower, prescription.targetRepsMax ?: lower)
        val priorLoad = previousSets.mapNotNull { it.loadKg }.maxOrNull()
        var workLoad = priorLoad ?: plan.targetLoadKg
        val compared = previousSets.take(prescription.targetSets)
        var workRepTargets = List(prescription.targetSets) { index ->
            compared.getOrNull(index)?.reps?.coerceIn(lower, upper) ?: lower
        }
        var duration = compared.mapNotNull { it.durationS }.maxOrNull() ?: plan.targetDurationS
        var reason = if (previous == null) {
            StrengthProgressionReason.FIRST_SESSION
        } else {
            StrengthProgressionReason.REPEAT_LOAD
        }
        when (plan.progression) {
            "double_progression" -> if (
                compared.size >= prescription.targetSets && compared.all { (it.reps ?: 0) >= upper }
            ) {
                if (workLoad != null) {
                    progressedLoad(workLoad, plan.loadStepKg)?.let {
                        workLoad = it
                        workRepTargets = List(prescription.targetSets) { lower }
                        reason = StrengthProgressionReason.REP_RANGE_ADVANCED
                    }
                } else {
                    val step = if (plan.repsPerSide) 2 else 1
                    val completedFloor = compared.mapNotNull { it.reps }.minOrNull() ?: upper
                    val advanced = min(
                        StrengthTrainingContract.MAX_REPS,
                        max(upper, completedFloor) + step,
                    )
                    workRepTargets = List(prescription.targetSets) { advanced }
                    if (advanced > completedFloor) {
                        reason = StrengthProgressionReason.BODYWEIGHT_REP_PROGRESS
                    }
                }
            } else if (compared.isNotEmpty()) {
                val step = if (plan.repsPerSide) 2 else 1
                workRepTargets = List(prescription.targetSets) { index ->
                    compared.getOrNull(index)?.reps
                        ?.let { min(upper, max(lower, it) + step) }
                        ?: lower
                }
            }
            "linear" -> {
                val current = workLoad
                if (
                    compared.size >= prescription.targetSets &&
                    compared.all { (it.reps ?: 0) >= lower } &&
                    current != null
                ) {
                    progressedLoad(current, plan.loadStepKg)?.let {
                        workLoad = it
                        reason = StrengthProgressionReason.LINEAR_ADVANCED
                    }
                }
            }
            "time" -> {
                val target = max(
                    plan.targetDurationS ?: 30,
                    compared.mapNotNull { it.durationS }.maxOrNull() ?: 0,
                )
                if (
                    compared.size >= prescription.targetSets &&
                    compared.all { (it.durationS ?: 0) >= target }
                ) {
                    val advanced = min(StrengthTrainingContract.MAX_DURATION_SECONDS, target + 5)
                    duration = advanced
                    if (advanced > target) reason = StrengthProgressionReason.TIME_ADVANCED
                } else {
                    duration = target
                }
            }
        }
        if (exercise.equipment == "bodyweight" && plan.targetLoadKg == null && priorLoad == null) {
            workLoad = null
        }
        val timed = plan.mode == "timed" || exercise.movementPattern == "cardio"
        val warmups = warmupRows(
            if (timed) 0 else plan.warmupSets,
            workLoad,
            workRepTargets.firstOrNull() ?: lower,
        )
        val work = (0 until prescription.targetSets).map { index ->
            StrengthPlannedSet(
                setType = if (exercise.equipment == "bodyweight" && workLoad == null) {
                    "bodyweight"
                } else {
                    "working"
                },
                reps = if (timed) null else workRepTargets[index],
                loadKg = workLoad,
                durationS = if (timed) duration ?: 30 else null,
                restSecondsAfter = null,
            )
        }.toMutableList()
        val clusterLoad = workLoad
        if (!timed && plan.setStyle == "drop" && clusterLoad != null) {
            val dropLoad = roundedLoad(clusterLoad * (1.0 - plan.dropPercent / 100.0))
            if (work.isNotEmpty() && dropLoad > 0.0 && dropLoad < clusterLoad) {
                work[work.lastIndex] = work.last().copy(restSecondsAfter = 0)
                work += StrengthPlannedSet(
                    setType = "drop",
                    reps = workRepTargets.lastOrNull() ?: lower,
                    loadKg = dropLoad,
                    durationS = null,
                    restSecondsAfter = null,
                )
            }
        } else if (!timed && plan.setStyle == "rest_pause" && work.isNotEmpty()) {
            work[work.lastIndex] = work.last().copy(restSecondsAfter = plan.restPauseSeconds)
            work += StrengthPlannedSet(
                setType = "rest_pause",
                reps = null,
                loadKg = clusterLoad,
                durationS = null,
                restSecondsAfter = null,
            )
        }
        return StrengthWorkoutPrescription(warmups + work, reason, previous?.session?.startedAt)
    }

    /**
     * Resolve the timer after a planned set. Warm-ups keep their prescription rest, while working
     * sets inside a superset wait only after the last movement in the round.
     */
    fun resolvedRestSeconds(
        target: StrengthPlannedSet,
        prescriptionRestSeconds: Int,
        continuesSuperset: Boolean,
    ): Int {
        if (target.setType == "warmup") return prescriptionRestSeconds
        return target.restSecondsAfter ?: if (continuesSuperset) 0 else prescriptionRestSeconds
    }

    fun estimatedOneRepMaximum(loadKg: Double, reps: Int): Double? {
        if (!loadKg.isFinite() || loadKg <= 0.0 || reps !in 1..10) return null
        return if (reps == 1) loadKg else loadKg * (1.0 + reps / 30.0)
    }

    fun bestEstimatedMaximum(
        exerciseId: String,
        sessions: List<StrengthSessionSnapshot>,
    ): StrengthEstimatedMaximum? =
        sessions.filter { it.session.endedAt != null }.flatMap { snapshot ->
            snapshot.sets.mapNotNull { set ->
                val load = set.loadKg
                val reps = set.reps
                if (
                    set.exerciseId != exerciseId || set.completedAt == null ||
                    set.setType !in setOf("working", "failure") || load == null || reps == null
                ) return@mapNotNull null
                val estimate = estimatedOneRepMaximum(load, reps) ?: return@mapNotNull null
                StrengthEstimatedMaximum(
                    exerciseId, estimate, load, reps, snapshot.session.id, snapshot.session.startedAt,
                )
            }
        }.maxWithOrNull(compareBy<StrengthEstimatedMaximum> { it.kilograms }.thenBy { it.recordedAt })

    private fun warmupRows(count: Int, workLoad: Double?, workReps: Int): List<StrengthPlannedSet> {
        if (count <= 0 || workLoad == null) return emptyList()
        val ladder = mapOf(
            1 to listOf(0.6 to 5),
            2 to listOf(0.5 to 5, 0.75 to 3),
            3 to listOf(0.4 to 5, 0.6 to 3, 0.8 to 2),
            4 to listOf(0.35 to 5, 0.5 to 4, 0.65 to 3, 0.8 to 2),
            5 to listOf(0.3 to 5, 0.45 to 4, 0.6 to 3, 0.72 to 2, 0.84 to 1),
        ).getValue(count.coerceAtMost(5))
        return ladder.mapNotNull { (loadFraction, suggestedReps) ->
            val load = roundedLoad(workLoad * loadFraction)
            if (load <= 0.0 || load >= workLoad) return@mapNotNull null
            StrengthPlannedSet(
                "warmup",
                max(1, min(workReps, suggestedReps)),
                load,
                null,
                null,
            )
        }
    }

    private fun roundedLoad(kilograms: Double): Double = floor(kilograms * 2.0 + 0.5) / 2.0

    private val PROGRESSION_SET_TYPES = setOf("working", "failure", "bodyweight")

    /**
     * ACSM recommends a 2–10% load increase after the target is exceeded. Respect the configured
     * increment while capping it at 10%; if 0.5 kg resolution would exceed that cap, hold the load.
     */
    private fun progressedLoad(current: Double, configuredStep: Double): Double? {
        if (!current.isFinite() || current <= 0.0 || !configuredStep.isFinite() || configuredStep <= 0.0) {
            return null
        }
        val maximumIncrease = current * 0.10
        val unrounded = current + min(configuredStep, maximumIncrease)
        var candidate = roundedLoad(unrounded)
        if (candidate - current > maximumIncrease + 0.000_001) {
            candidate = floor(unrounded * 2.0) / 2.0
        }
        return candidate.takeIf { it > current }
    }

}
